lib/segment/src/vector_storage/quantized/quantized_multivector_storage.rs

```diff
@@ -0,0 +1,180 @@
+            offsets,
+            multi_vector_config: *multi_vector_config,
+            encoded_query: PhantomData,
+        })
+    }
+
+    fn score_point_max_similarity(&self, query: &Vec<TEncodedQuery>, vector_index: u32) -> f32 {
+        let vectors_count = self.offsets[vector_index as usize].count;
+        let vectors_offset = self.offsets[vector_index as usize].offset;
+        let mut sum = 0.0;
+        for inner_query in query {
+            let mut max_sim = ScoreType::NEG_INFINITY;
+            // manual `max_by` for performance
+            for i in 0..vectors_count {
+                let sim = self
+                    .quantized_storage
+                    .score_point(inner_query, vectors_offset + i);
+                if sim > max_sim {
+                    max_sim = sim;
+                }
+            }
+            // sum of max similarity
+            sum += max_sim;
+        }
+        sum
+    }
+
+    fn score_internal_max_similarity(&self, vector_a_index: u32, vector_b_index: u32) -> f32 {
+        let vector_a_count = self.offsets[vector_a_index as usize].count;
+        let vector_b_count = self.offsets[vector_b_index as usize].count;
+        let vector_a_offset = self.offsets[vector_a_index as usize].offset;
+        let vector_b_offset = self.offsets[vector_b_index as usize].offset;
+        let mut sum = 0.0;
+        for a in 0..vector_a_count {
+            let mut max_sim = ScoreType::NEG_INFINITY;
+            // manual `max_by` for performance
+            for b in 0..vector_b_count {
+                let sim = self
+                    .quantized_storage
+                    .score_internal(vector_a_offset + a, vector_b_offset + b);
+                if sim > max_sim {
+                    max_sim = sim;
+                }
+            }
+            // sum of max similarity
+            sum += max_sim;
+        }
+        sum
+    }
+}
+
+impl<TEncodedQuery, QuantizedStorage> EncodedVectors<Vec<TEncodedQuery>>
+    for QuantizedMultivectorStorage<TEncodedQuery, QuantizedStorage>
+where
+    TEncodedQuery: Sized,
+    QuantizedStorage: EncodedVectors<TEncodedQuery>,
+{
+    // TODO(colbert): refactor `EncodedVectors` to support multi vector storage after quantization migration
+    fn save(&self, _data_path: &Path, _meta_path: &Path) -> std::io::Result<()> {
+        unreachable!("multivector quantized storage should be saved using `self.save_multi` method")
+    }
+
+    // TODO(colbert): refactor `EncodedVectors` to support multi vector storage after quantization migration
+    fn load(
+        _data_path: &Path,
+        _meta_path: &Path,
+        _vector_parameters: &quantization::VectorParameters,
+    ) -> std::io::Result<Self> {
+        unreachable!(
+            "multivector quantized storage should be loaded using `self.load_multi` method"
+        )
+    }
+
+    fn encode_query(&self, query: &[f32]) -> Vec<TEncodedQuery> {
```

Comment:
`EncodedVectors` wasn't designed for multivectors, `query` here is a flattened data. Like with save/load, it requires refactoring